ros2/src/autorccar_costmap/src/costmap.cpp

```diff
@@ -0,0 +1,227 @@
+    global_center_y = global_size_y / 2;
+    InitCostmap(global_costmap, global_size_x, global_size_y);
+
+    // Initialize local costmap
+    local_size_x = (int)local_width / resolution;
+    local_size_y = (int)local_height / resolution;
+    local_center_x = local_size_x / 2;
+    local_center_y = local_size_y / 2;
+    InitCostmap(local_costmap, local_size_x, local_size_y);
+
+    // Allocate memory
+    cur_point_cloud.reset(new pcl::PointCloud<pcl::PointXYZI>());
+    cur_point_cloud_transformed.reset(new pcl::PointCloud<pcl::PointXYZI>());
+    cur_point_cloud_filtered.reset(new pcl::PointCloud<pcl::PointXYZI>());
+}
+
+void Costmap::LoadConfig(const std::string& config) {
+    std::ifstream fin(config);
+    if (!fin) {
+        std::cerr << "Failed to open file: " << config << std::endl;
+        exit(-1);
+    }
+    fin.close();
+
+    YAML::Node cfg = YAML::LoadFile(config);
+
+    resolution = cfg["global"]["resolution"].as<double>();
+    cnt_limit = cfg["global"]["updateEveryNthLidar"].as<unsigned int>();
+
+    global_width = cfg["global"]["width"].as<int>();
+    global_height = cfg["global"]["height"].as<int>();
+
+    local_width = cfg["local"]["width"].as<int>();
+    local_height = cfg["local"]["height"].as<int>();
+
+    std::cout << "Successfully loaded the config" << std::endl;
+}
+
+void Costmap::InitCostmap(Eigen::MatrixXd& costmap, int size_x, int size_y) {
+    costmap.resize(size_x, size_y);
+    costmap.setConstant(ProbabilityToLogOdds(P_UNKNOWN));
+}
+
+void Costmap::UpdatePointCloud(const pcl::PointCloud<pcl::PointXYZI>::Ptr pcl) {
+    cur_point_cloud->clear();
+    cur_point_cloud_transformed->clear();
+    cur_point_cloud_filtered->clear();
+
+    pcl::copyPointCloud(*pcl, *cur_point_cloud);
+
+    UpdateCostmapFlag();
+}
+
+void Costmap::UpdatePose(Eigen::Matrix<double, 4, 4>& trans) { cur_pose = trans; }
+
+void Costmap::UpdateCostmapFlag() {
+    cnt_iter += 1;
+
+    if (cnt_iter >= cnt_limit) {
+        costmap_flag = true;
+        cnt_iter = 0;
+    }
+}
+
+void Costmap::UpdateCostmap() {
+    // Get cell indices of current pose. This will be used as the start point of the bresenham line.
+    int sx = floor(cur_pose(0, 3) / resolution) + global_center_x;  // assume (global_center_x, global_center_y) = (0,0)
+    int sy = floor(cur_pose(1, 3) / resolution) + global_center_y;
+
+    // Transform point cloud from body to world frame.
+    pcl::transformPointCloud(*cur_point_cloud, *cur_point_cloud_transformed, cur_pose);
+
+    // Filtering
+    pcl::PassThrough<pcl::PointXYZI> zFilter;
+    zFilter.setInputCloud(cur_point_cloud_transformed);
+    zFilter.setFilterFieldName("z");
+    zFilter.setFilterLimits(0.0, 1.0);
+    zFilter.filter(*cur_point_cloud_filtered);
+
+    for (const auto& pt : *cur_point_cloud_filtered) {
+        // Get cell indices of the pointcloud
+        int x = floor(pt.x / resolution) + global_center_x;
+        int y = floor(pt.y / resolution) + global_center_y;
```

Comment:
```suggestion
        int x = std::floor(pt.x / resolution) + global_center_x;
        int y = std::floor(pt.y / resolution) + global_center_y;
```

STL 함수를 쓸 땐 `std::`를 붙이는게 나중에 혹시모를 충돌을 방지해줍니다.

는 다른데는 다 붙고 여기만 빠진것같네 ㅋㅋ :joy: